Fix confusing output at the end of the simulation

MarcelWolkober opened the pull request:
## Main changes of this PR

This PR fixes the confusing output of preCICE, which shows that the simulation has finished.
The output when the simulation is finished is now more understandable.

For example, the end of the output of the Oscillator tutorial is now:

```
... Reached end at: final time-window: 100, final time: 1
```

At the end of the PR are the logs of all the variances of serial/parallel and implicit/explicit, run with the solver dummies.

## Motivation and additional information

<!--
Short rational why preCICE needs this change. If this is already described in an issue a link to that issue (closes #123) is sufficient.
-->
Closes issue #2101.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I stuck to C++17 features.
* [x] I stuck to CMake version 3.22.1.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

<!-- Tag people next to each point and add points for specific questions -->

Hi @fsimonis, 
this is how I would change the output, which is basically your second proposed scheme.
Do you want any additional changes, for example, that the `_maxTimeWindows` is also included if it is provided?

I am a little confused about ~~the `make changelog` because it says I need a PR first and also~~ what kind of test I should create. 

Edit: I've added a second commit with the new changelog file, I hope this is correct.

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

<!-- add more questions/tasks if necessary -->


## Solver Dummy Logs (short)

### Serial implicit

SolverOne:

```
[...]
preCICE: iteration: 3 of 3 (min 2), time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: no, read-iteration-checkpoint
DUMMY: Reading iteration checkpoint
preCICE: Time window completed
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

SolverTwo:

```
[...]
preCICE: iteration: 3 of 3 (min 2), time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: no, read-iteration-checkpoint
DUMMY: Reading iteration checkpoint
preCICE: Mapping "Data-Two" for t=2 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: No converge measures defined.
preCICE: Time window completed
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

### Serial explicit

SolverOne:

```
[...]
preCICE: iteration: 1, time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: yes
DUMMY: Advancing in time
preCICE: Time window completed
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

SolverTwo:

```
[...]
preCICE: iteration: 1, time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: yes
DUMMY: Advancing in time
preCICE: Mapping "Data-Two" for t=2 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Time window completed
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

### Parallel implicit

SolverOne:

```
[...]
preCICE: iteration: 3 of 3 (min 2), time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: no, read-iteration-checkpoint
DUMMY: Reading iteration checkpoint
preCICE: Time window completed
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

SolverTwo:

```
[...]
preCICE: iteration: 3 of 3 (min 2), time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: no, read-iteration-checkpoint
DUMMY: Reading iteration checkpoint
preCICE: Mapping "Data-Two" for t=2 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: No converge measures defined.
preCICE: Time window completed
preCICE: Mapping "Data-One" for t=2 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

### Parallel explicit

SolverOne:

```
[...]
preCICE: iteration: 1, time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: yes
DUMMY: Advancing in time
preCICE: Time window completed
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

SolverTwo:

```
[...]
preCICE: iteration: 1, time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: yes
DUMMY: Advancing in time
preCICE: Mapping "Data-Two" for t=2 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Time window completed
preCICE: Mapping "Data-One" for t=2 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: Reached end at: final time-window: 2, final time: 2
[...]
```

## Example Composition Coupling Logs

```
[...]
199: 13:02:02.086812|IB|0|impl::ParticipantImpl|l431|advance|EB: iteration: 1, time-window: 2 of 2, time: 2, time-window-size: 1, max-time-step-size: -0, ongoing: yes, time-window-complete: yes
199: IA: iteration: 2 of 2 (min 2), time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: no, read-iteration-checkpoint
--
199: 13:02:02.089375|IB|0|impl::ParticipantImpl|l431|advance|EB: Reached end at: final time-window: 2, final time: 2
199: IA: Reached end at: final time-window: 2, final time: 2
```

MarcelWolkober commented:
After talking to @uekerman I added now all the different ways to run the coupling scheme with the solver dummy. So we can use this PR like a manually created testing docu. I can also paste the whole logs somewhere (else) if that is wanted. Should I run/test anything else or restructure the logs? Otherwise, I'd like to change this draft into an official PR soon.

fsimonis commented:
Hi there, thanks for the contribution and the logs!

Currently, the log shows the number of the time window that would have started if the simulator went on.
I think it would be easier to understand if the output would state the number of the final time-window and also refer to it as "final" or "last".

Also, the iteration output of implicit schemes will always say "iteration 1" when the coupling stops. It would be nice if this wouldn't show when the coupling isn't ongoing anymore.

MarcelWolkober commented:
Hi @fsimonis, thanks for the feedback!
 
> Also, the iteration output of implicit schemes will always say "iteration 1" when the coupling stops. It would be nice if this wouldn't show when the coupling isn't ongoing anymore.

Do you mean something like this? :

```console
preCICE: iteration: 3 of 3 (min 2), time-window: 2 of 2, time: 1, time-window-size: 1, max-time-step-size: 1, ongoing: yes, time-window-complete: no, read-iteration-checkpoint
DUMMY: Reading iteration checkpoint
preCICE: Mapping "Data-Two" for t=2 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: No converge measures defined.
preCICE: Time window completed
preCICE: Mapping "Data-One" for t=2 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: iteration: 1 of 3 (min 2), Reached end at: final time-window: 2, final time: 2,
```

so should I not print out the "iteration: 1 of 3 (min 2)," part? Or should I try to fix the output? 

If I don't print it, I will not print it for all types of coupleling schemes, do you agree?

MarcelWolkober commented:
I've updated the code and the logs in the initial PR message with the new format where the "iteration: 1 of 3 (min 2)," is not written, I hope this was okay to do without confirmation.

Additionally, I added a small check if an action state is not empty to decide whether it should print a comma. 

E.g., the change:

```
preCICE: iteration: 1, Reached end at: time-window: 3, time: 2,
```

to 

```
preCICE: Reached end at: final time-window: 2, final time: 2
```

Please let me know if there should be any additional changes :)

fsimonis commented:
As a final step, please have a look on how this reflects in compositional coupling schemes. You may want to tweak the formatting to make this easier to read.

This test should give you a good indication.
```
ctest -VV -R precice.Integration/Serial/Compositional/Data/ImplicitFirst/AllParallel
```

You are looking for participant IB, who is coupled to both IA and EB, handling two coupling schemes simultaneously.

There are two lines in the output, the first one stating with `12:39:55.663140|IB|0|impl::ParticipantImpl|l437|advance`.

You can use grep with (A)fter context of one line to make this less overwhelming.

```
ctest -R precice.Integration/Serial/Compositional/Data/ImplicitFirst/AllParallel -VV | grep "IB|.*|advance" -A1
```

Here is an example of vanilla preCICE

```
201: 12:39:55.663140|IB|0|impl::ParticipantImpl|l437|advance|EB: iteration: 1, time-window: 3 of 2, time: 2, time-window-size: 1, max-time-step-size: 0, ongoing: no, time-window-complete: yes,
201: IA: iteration: 1 of 2 (min 2), time-window: 3 of 2, time: 2, time-window-size: 1, max-time-step-size: 0, ongoing: no, time-window-complete: yes,
```

fsimonis reviewed (approved):
Looking very good now!

Thanks for the contribution!

fsimonis commented on src/cplscheme/BaseCouplingScheme.cpp:
```diff
@@ -402,7 +402,7 @@ void BaseCouplingScheme::secondExchange()
         }
       }
-      //update iterations
+      // update iterations
```
No need. I can squash and merge.
